lib/thesis/notifications.ex

```diff
@@ -0,0 +1,163 @@
+    :add_page,
+    :import_export_restore,
+    :other
+  ]
+
+  def all(page) do
+    %{
+      "page-settings" => [],
+      "add-page" => [],
+      "import-export-restore" => [],
+      "other" => []
+    }
+    |> notifications_from_host_app(notifications())
+    |> notifications_regarding_env(Mix.env)
+    |> notifications_regarding_page(page)
+  end
+
+  def notifications_regarding_page(acc, nil), do: acc
+  def notifications_regarding_page(acc, %_{}) do
+    acc
+  end
+
+  def notifications_regarding_env(acc, :test), do: acc
+  def notifications_regarding_env(acc, _) do
+    acc
+    |> notification_to_run_migration_for_version_0_2_1()
+  end
+
+  @doc """
+  Adds notifications provided by the host app to the accumulator.
+
+      iex> notifications_from_host_app(%{"page-settings" => []}, [page_settings: ["notif"]])
+      %{"page-settings" => ["notif"]}
+      iex> notifications_from_host_app(%{"page-settings" => []}, [page_settings: :invalid])
+      %{"page-settings" => []}
+      iex> notifications_from_host_app(%{"page-settings" => ["notif1"]}, [page_settings: ["notif2"]])
+      %{"page-settings" => ["notif2", "notif1"]}
+      iex> notifications_from_host_app(%{"other" => [], "add-page" => ["notif"]}, [other: ["notif"]])
+      %{"other" => ["notif"], "add-page" => ["notif"]}
+      iex> notifications_from_host_app(%{"other" => []}, nil)
+      %{"other" => []}
+  """
+  def notifications_from_host_app(acc, notifications) when is_list(notifications) do
+    case verify_notifications_structure(notifications) do
+      true ->
+        Enum.reduce(notifications, acc, fn({type, list}, acc) ->
+          update_notifications_map(acc, type, list)
+        end)
+      _ -> acc
+    end
+  end
+  def notifications_from_host_app(acc, _), do: acc
+
+  @doc """
+  Verifies that the notifications provided by the host app are in the correct format.
+
+      iex> verify_notifications_structure([:invalid])
+      false
+      iex> verify_notifications_structure(%{})
+      false
+      iex> verify_notifications_structure(%{invalid_key: []})
+      false
+      iex> verify_notifications_structure(%{add_page: "invalid value"})
+      false
+      iex> verify_notifications_structure(%{add_page: ["notif", "notif", :atom]})
+      false
+      iex> verify_notifications_structure([add_page: [], page_settings: ["notif"]])
+      false
+  """
+  def verify_notifications_structure(notifications) when is_list(notifications) and length(notifications) == 0, do: false
+  def verify_notifications_structure(notifications) when is_list(notifications) do
+    case Keyword.keyword?(notifications) do
+      true ->
+        notifications
+        |> Enum.map(fn(n) ->
+          n
+          |> is_notification_key_allowed?()
+          |> is_notification_value_data_type_valid?()
+          |> is_notification_value_valid?()
+        end)
+        |> Enum.all?()
+      _ -> false
+    end
+  end
+  def verify_notifications_structure(_), do: false
```

Comment:
Here we run some checks for the notifications that are provided to us from the host app. We check that the notification key (type) is allowed, then we check if the notification is a List or Function. Lastly, we check that the items in the List are Strings.